### ISSUE TITLE
[hdpowerview] Shade secondary position calculated differently when getting vs setting

### ISSUE THREAD

**Issue by arroyoj** [openhab/openhab-addons#11543]
## Expected Behavior
When setting the position of the secondary rail of a top-down/bottom-up Powerview shade, the secondary rail should move to the set position and the OpenHAB state of the secondary rail should consistently reflect that position.

## Current Behavior
After setting the position of the secondary rail using a PercentType command, the rail moves properly to the correct position, and the OpenHAB state is updated appropriately. However, after the add-on refreshes the position from the raw internal Powerview value, the OpenHAB state is incorrectly updated to the the "inverted" value within the 0-100 range (ie, 0 become 100, 70 becomes 30).

Here is an example from my logs after sending a secondary rail (with Item name `LivingRoom_Shade1_Top`) the command to move to a position value of 70:
```
2021-11-07 21:46:28.100 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'LivingRoom_Shade1_Top' received command 70
2021-11-07 21:46:28.101 [INFO ] [penhab.event.ItemStatePredictedEvent] - Item 'LivingRoom_Shade1_Top' predicted to become 70
2021-11-07 21:46:28.102 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'LivingRoom_Shade1_Top' changed from 0 to 70
2021-11-07 21:47:17.739 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'LivingRoom_Shade1_Top' changed from 70 to 30
```
The state is initially set to the correct value of 70, and the shade moves appropriately, but then after the add-on refreshes the state of all of the shades, the value is incorrectly set to 30.

## Possible Solution
First of all, thanks to the team for the HD Powerview add-on and especially @andrewfg for adding support for the dual action shades. Looking at the code, it appears that the set and get functions that convert between OpenHAB coordinates and Powerview coordinates are not reciprocal. 

Here is the code for the `setPosition2` function, in which setting an input PercentType value of 100 for the secondary rail will set the rail to the Powerview `MAX_SHADE` value and move the rail to the bottom (see line 214):
https://github.com/openhab/openhab-addons/blob/8dd4559443422ea9e034ea480a425154a7292cd5/bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/api/ShadePosition.java#L201-L219

Here is the code for the corresponding `getPosition2` function, which improperly subtracts the calculated value from 100, thereby inverting it compared to the set function (ie, the `MAX_SHADE` value is converted to a PercentType of 0, rather than the correct value of 100) (see line 239):
https://github.com/openhab/openhab-addons/blob/8dd4559443422ea9e034ea480a425154a7292cd5/bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/api/ShadePosition.java#L221-L244

I believe the calculation in the `setPosition2` function is correct, such that both the primary and secondary rails use an OpenHAB value of 0 for the top of the window and 100 for the bottom of the window. In that case, I think the `getPosition2` function just needs to have the "100 - " removed from line 239 above.

## Steps to Reproduce (for Bugs)
1. Send a PercentType command to an item linked to the secondary channel of a Powerview shade
2. Wait for OpenHAB to refresh state of shades from hub
3. Secondary rail position will change to 100 - [original command value]

## Context
I discovered this problem when trying to use a Slider UI to control my shades instead of the standard Up/Down/Stop roller shutter UI. With the slider UI for the secondary rail, I would set a value, and then the state would jump to a different value shortly after being set.

## Your Environment
* Version used: OpenHAB 3.1.0; HD PowerView Binding 3.1.0

**Comment by andrewfg** [comment]
@arroyoj  thank-you for the feedback; I will have a look at it in the next couple of days.

**Comment by andrewfg** [comment]
@arroyoj yes I think your assertion is correct, and I have made a potential fix [here](https://github.com/andrewfg/openhab-addons/tree/hdpowerview). However since I do not have a physical device for doing actual tests, I need you to check the solution on your system and if it is OK then I can formally open a Pull Request.

=> So can you please download the new [JAR file HERE](https://github.com/andrewfg/openhab-addons/blob/hdpowerview/bundles/org.openhab.binding.hdpowerview/org.openhab.binding.hdpowerview-3.2.0-SNAPSHOT.jar) and put it in your addons folder, and confirm that it is indeed working correctly for you? (And that nothing else was broken..)

**Comment by arroyoj** [comment]
@andrewfg thank you for putting together the updated JAR file. I tested it out, and it worked perfectly. I tested sending PercentType commands to both the primary and secondary rails, as well as sending Up/Down/Stop commands to both rails. In all cases, the shades moved correctly. The secondary rail position value also stayed consistent after it was set and did not flip to the "inverted" value.

While testing, I did come across another bug with the Powerview hub not reporting the secondary rail position after setting the primary rail, but I need to do a bit more testing and then I'll file a separate issue.

Thanks so much for the quick fix!

**Comment by andrewfg** [comment]
@arroyoj many thanks for reporting the issue and for testing the solution. 
Respective PR #11552 has been opened.

**Comment by jlaur** [comment]
@andrewfg, @arroyoj - just for my understanding and verifying documented behaviour vs. actual/intended behaviour with PR change: I also have a top down/bottom up Duette shade. When going from fully opened (both at top) to approximately this:
![image](https://user-images.githubusercontent.com/19519842/141323884-52477f70-2d29-4e8e-91f4-f8ae449e49fa.png)

I receive these values (with main branch, i.e. without PR fix):
2021-11-11 16:22:45.841 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Blind11_Position' changed from 1 to 87
2021-11-11 16:22:45.844 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Blind11_Secondary' changed from 99 to 21

Is this wrong or correct according to new behaviour and README? I'm actually not sure how to understand the README right now for top down/bottom up. Maybe even more clarification is needed here, or am I just stupid? :-)

**Comment by arroyoj** [comment]
@jlaur, great question, and I find this very confusing, too. When I was looking at the comments in the code above, it seemed like the set function was behaving as expected: "Secondary, upper, top-down rail of a dual action shade. Uses a coordinate system that is NOT inverted in relation to OpenHAB". That would also mean the primary and secondary rails are moving in the same OpenHAB coordinate system (top 0, bottom 100), which seemed "right" to me. However, now that I'm looking at the README after reading your comment, I'm not sure what is the "right" behavior for these shades, and I'm not sure the README is consistent with the setter code. In the README, it seems to say that for the secondary rail, sending the UP command would move the rail *down* while setting the position to 0%:
![image](https://user-images.githubusercontent.com/82246664/141392477-b6258204-454f-4b39-9354-b72823d5d3db.png)
However, with v3.1.0 (before the PR fix), when I send my secondary rail the UP command, it moves the secondary rail *up* (which seems intuitive) and sets it to 100%, before it then flips to returning a value of 0%. I think UP should make the rail move up, because otherwise that's too confusing, but I don't know whether that should set it to 100% or 0%. To me, it makes more sense for the top of the window to always be 0% for both rails and the bottom to always be 100% for both rails, but I'm fine with different behavior as long as set and get operations are consistent with each other. But, now that you mention it, it does seem like the original set behavior (and the new get behavior) is not the same as what is documented in the README. @andrewfg, what do you think of the behavior vs the documentation?

@jlaur, I was also just wondering, in the log snippet that you posted, did you move the two rails individually, or did you trigger a scene to move them? I've found that with triggering scenes, there is not the "jump" in OpenHAB state where it flips between set and get, because the hub sets the rails directly. The get/set bug seems to only show up when the secondary rail is moved manually through OpenHAB.

**Comment by andrewfg** [comment]
^
I wrote the ReadMe based on documentation provided by Hunter Douglas; I will have a look tomorrow to see if I can find that again. But, as I recall, there are a number of issues..
- They use primary and secondary positions in different ways for different types of shades (e.g. Venetian blinds, top down bottom up shades, left right rotating slat shades etc.) so the binding code has to consistently handle all scenarios
- As a general rule though, the primary position uses a coordinate system that is ABSOLUTE (usually the top or the left of the actual window opening), but the secondary position uses a coordinate system that is RELATIVE to the primary I.e an OFFSET to the primary. So in a Venetian blind the primary position moves the bottom rail up and down, and the secondary position adds a twist to the slats relative to the primary. And in a top down bottom up shade the primary position moves the position of the bottom rail up or down in the physical window, and the secondary position moves the position of the top rail relative to the bottom rail.
- When considering the meaning of ‘UP’ and ‘DOWN’ in relation to OpenHAB, the primary position usually maps to physically open or physically closed, which may map to bottom rail up, or left/right shade to the left etc. But what does ‘UP’ mean for the secondary position in relation to the tilt of the slats on a Venetian blind?? ..
- You also need to consider the relationship between ‘UP’ and ‘UP ARROW’ on the HD remote control. Not to mention ‘UP SWIPE’ in the app.

I will post the HD source document when I find it.

**Comment by jlaur** [comment]
@arroyoj - quick reply for now: I moved the rails individually using the PowerView app to produce that log snippet. So only retrieving the values from the hub.

**Comment by andrewfg** [comment]
> I receive these values (with main branch, i.e. without PR fix):
> 2021-11-11 16:22:45.841 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Blind11_Position' changed from 1 to 87
> 2021-11-11 16:22:45.844 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Blind11_Secondary' changed from 99 to 21
> Is this wrong or correct according to new behaviour and README?

IMHO this is correct. To be specific..
- at the start of the operation: the bottom rail is at 1% down relative to the window frame top, and the top rail is at 99% up relative to the bottom rail (albeit that 0% .. 100% are notionally indistinguishable since the physical distance between them is zero)
- at the end of the operation: the bottom rail has moved down to 87% from the top of the window frame, and the top rail has moved to 21% up from the bottom rail

**Comment by arroyoj** [comment]
@andrewfg, thanks for that clarification about how to interpret the values in @jlaur example. I misunderstood how the secondary rail values were supposed to be reported. If that is the expected behavior, then the original getter function was "right" (in that example, only the OpenHAB getter was being used), and the original setter function would need to change. However, just inverting the setter function would also invert the current UP/DOWN commands, since UP sets the secondary rail to 0 here (same as when UP is sent to the primary rail): https://github.com/openhab/openhab-addons/blob/8dd4559443422ea9e034ea480a425154a7292cd5/bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewShadeHandler.java#L127

**Comment by andrewfg** [comment]
At least the latest version changes make the Setter and Getter self consistent. But I think I should have changed the Setter rather than changing the Getter. In HD speak, UP=OPEN rather than UP='physically up' -- this reflects the buttons of the remote control where the upwards pointing triangle symbol means OPEN, and the downwards pointing triangle means CLOSE.

**Comment by arroyoj** [comment]
@andrewfg, first of all, thanks again for getting the secondary rail functionality into OpenHAB and addressing this bug so quickly. I'm sorry I got the setter and getter behavior confused, but I'm glad they are at least self-consistent now.

In terms of the HD remote, it has two buttons (left/down and right/up) that control the secondary rail, and those always move the secondary rail in the expected way (down moves the rail down). I thought the OpenHAB secondary rail UI would work like those buttons, and currently they do :). The HD remote also seems to overload the primary rail buttons (Up/Open and Down/Close) like you said, such that they usually only control the primary rail but in some cases also move the secondary rail in the "opposite" way to make the shade fully open or closed. Personally, I found that behavior confusing, and since the OpenHAB shades UI just shows up and down arrows, I didn't realize the primary channel's overloaded behavior was being applied to the secondary channel.

Since the original getter/setter bug is fixed, but the behavior vs README vs physical moving is still a bit confusing, I'll open a new issue where we can discuss and hopefully resolve this. Thanks!

**Comment by andrewfg** [comment]
> the behavior vs README vs physical moving is still a bit confusing

Yeah. And the HD API description (below) does not really make it any clearer...

![image](https://user-images.githubusercontent.com/893994/141644005-997d1cbd-6232-46fc-801d-f136f356d8ab.png)

It is complicated by the fact that I don't actually own a top down bottom up shade, so I cannot code-test-debug it directly by myself.

Therefore I think you are right that we might need to open another Issue resp. PR in order to fully resolve this. But I will need you both @jlaur and @arroyoj to help with the testing part of the cycle. => I hope that is Ok with you?

Apropos the ▲▼ buttons, it is important to keep in mind that in HD language (say) the ▲ button always means `OPEN` but -- **_depending on the type of shade_** -- might also mean the rail is physically either `UP` or `DOWN`!

![image](https://user-images.githubusercontent.com/893994/141644342-53298181-2505-4bc2-b3a2-b880a1bbc474.png)

PS I suspect that we might also need to look at `PosKind` again as well, since for Venetian blinds with rotating vanes, the `Position` value has yet another meaning entirely.

![image](https://user-images.githubusercontent.com/893994/141644479-57801818-1e48-4e3a-9706-74755eaa43d5.png)

**Comment by jlaur** [comment]
@andrewfg - I'm a bit behind, but I'll read everything that's been written and try to catch up. I'll of course also help verifying and testing. I have so far only read values from the secondary channel, so haven't been affected by the inconsistency with writing values or given much thought about up/down behaviour.

Currently only have this rule which is affected:

```
rule "Duette in walk-in down when bath detected"
when
    Item Netatmo_Bathroom_Humidity changed
then
    var previousHumidity = (Netatmo_Bathroom_Humidity.historicState(now.minusMinutes(12)).state as Number).doubleValue
    var currentHumidity = (Netatmo_Bathroom_Humidity.state as Number).doubleValue
    var deltaHumidity = currentHumidity - previousHumidity
    if (deltaHumidity >= 4 && (Blind11_Position.state < 99 || Blind11_Secondary.state < 51))
    {
        PowerViewHub_Scene_Private.sendCommand(ON)
    }
end
```

In this case I've been using a scene, thus the hub has been managing the positions. Logic: If bottom rail not entirely down or top rail less than halfway up from bottom, trigger scene.

**Comment by jlaur** [comment]
I can add that the Pebble remote for Duette top down/bottom works like this:
- **Open (with arrow up):** Moves lower rail up towards the position of the upper rail. So if upper rail is at topmost position and lower rail it at bottommost position, the shade is essentially closed i.e. whole windows is covered. So lower rail will be meet upper rail at the top, and the shade will be fully open. If the upper rail is not on top, the lower rail will stop when it meets upper rail.
- **Close (with arrow down):** Moves lower rail towards bottom.
- **Right vane button (with arrow up):** Moves upper rail towards top.
- **Left vane button (with arrow down):** Moves upper rail down towards the position of the lower rail. When it meets the lower rail, it will stop.

This is just to explain the Hunter Douglas Pebble behaviour with top down/bottom up shades.

**Comment by jlaur** [comment]
Wow, also a lot of reading in the original issue here: #4351.

**Comment by arroyoj** [comment]
This sounds great. I'll open the new issue in the next few days, linking to this one and the original one for reference. I'm also happy to help with testing.

**Comment by andrewfg** [comment]
^
Thanks to you both for your support.

> the Pebble remote for Duette top down/bottom works like this

In order to assess if there is indeed a (future) Issue that needs to be opened, (or just an update of the ReadMe file), it would be helpful to have a log.trace of the JSON payload for position2 REST API calls for all of the following cases (using the fixed JAR from my last merged PR)..
1. The GetPosition2 payload when each of your above Pebble Remote button commands are executed.
2. When the Channel is linked to a PercentType slider Item in OH: the SetPosition2 payload, GetPosition2 payload, and physical position when the slider 0%, and 100% commands are sent.
3. When the Channel is linked to an OnOffType switch Item in OH: ditto as above, when the switch OFF, and ON commands are sent.
4. When the Channel is linked to a RollerShutter Item in OH: ditto as above, when the roller shutter UP, DOWN, and STOP commands are sent.

**Comment by arroyoj** [comment]
@andrewfg, I just opened #11593 to continue this discussion and try to resolve some of these differences in behavior. I think it would be great if we could first come to an agreement on what the behavior should be, and then I'm happy to help test and submit the log traces that you asked for above. Hopefully the tables that I included in #11593 helps to address some of your questions about how the Remote behaves with dual-rail shades.
